nncf/quantization/algorithms/layerwise/scheduler.py

```diff
@@ -0,0 +1,237 @@
+# Copyright (c) 2024 Intel Corporation
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#      http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from copy import deepcopy
+from dataclasses import dataclass
+from dataclasses import field
+from typing import Dict, List, NamedTuple, Set, TypeVar
+
+from nncf.common.graph.graph import NNCFGraph
+from nncf.common.graph.graph import NNCFNode
+from nncf.quantization.passes import transform_to_inference_graph
+
+TModel = TypeVar("TModel")
+
+
+class NodeOutputPort(NamedTuple):
+    """
+    Represents the output port of a node in the graph.
+
+    :param node_name: The name of the node.
+    :param output_port: The output port index.
+    """
+
+    node_name: str
+    output_port: int
+
+
+@dataclass
+class LayerwiseStep:
+    """
+    Represents a single step in the layer-wise processing of the graph.
+
+    :param target_node_map: Mapping of target nodes to their inputs/outputs ports
+        and corresponding output ports in the subgraph.
+    :param subgraph_inputs: List of input ports for the subgraph.
+    :param subgraph_outputs: List of output ports for the subgraph.
+    """
+
+    target_node_map: Dict[NNCFNode, Dict[int, NodeOutputPort]]
+    subgraph_inputs: List[NodeOutputPort]
+    subgraph_outputs: List[NodeOutputPort]
+
+
+@dataclass
+class SimplePath:
+    """
+    Represents a simple path in the graph with input and output nodes.
+
+    :param input_nodes: Set of input nodes.
+    :param output_nodes: Set of output nodes.
+    :param inputs: Set of input ports.
+    """
+
+    input_nodes: Set[NNCFNode]
+    output_nodes: Set[NNCFNode]
+    inputs: Set[NodeOutputPort] = field(default_factory=set)
+
+
+class LayerwiseScheduler:
+    """
+    Scheduler that determines the order of layer-wise steps for processing a graph based on a given strategy.
+    """
+
+    def __init__(self, add_additional_outputs: bool = False):
+        """
+        :param strategy: The strategy to use for scheduling.
+        """
+        self.add_additional_outputs = add_additional_outputs
+
+    def schedule(
```

Comment:
Maybe it's worth adding some description of graph traverse and resulting paths. How this procedure handles residuals and so on.

Reply:
done